src/frequenz/sdk/timeseries/periodic_feature_extractor.py

```diff
@@ -0,0 +1,392 @@
+            )
+
+        log.debug("Initializing PeriodicFeatureExtractor!")
+        log.debug(
+            "Period between two succeeding intervals (in samples): %i",
+            self._period,
+        )
+
+    def _timestamp_to_rel_index(self, timestamp: datetime) -> int:
+        """
+        Get the index of a timestamp relative to the oldest sample in the MovingWindow.
+
+        In other word consider an integer axis where the zero is defined as the
+        oldest element in the MovingWindow. This function returns the index of
+        the given timestamp an this axis.
+
+        Args:
+            timestamp: A timestamp that we want to shift into the window.
+
+        Returns:
+            The index of the timestamp shifted into the MovingWindow.
+        """
+        # align timestamp to the sampling period
+        timestamp = self._buffer.normalize_timestamp(timestamp)
+
+        # distance between the input ts and the ts of oldest know samples (in samples)
+        dist_to_oldest = int(
+            abs((timestamp - self._moving_window.time_bound_oldest).total_seconds())
+            / self._sampling_period
+        )
+
+        log.debug("Shifting ts: %s", timestamp)
+        log.debug(
+            "Oldest timestamp in buffer: %s", self._moving_window.time_bound_oldest
+        )
+        log.debug("Distance to the oldest sample: %i", dist_to_oldest)
+
+        return dist_to_oldest
+
+    def _num_windows(
+        self, window: NDArray[np.float_] | MovingWindow, window_size: int
+    ) -> int:
+        """
+        Get the number of windows that are fully contained in the MovingWindow.
+
+        This method calculates how often a given window, defined by it size, is
+        fully contained in the MovingWindow or any numpy ndarray given the
+        period between two window neighbors.
+
+        Args:
+            window: The buffer that is used for the average calculation.
+            window_size: The size of the window in samples.
+
+        Returns:
+            The number of windows that are fully contained in the MovingWindow.
+        """
+        num_windows = math.floor(len(window) / self._period)
+        if len(window) - num_windows * self._period >= window_size:
+            num_windows += 1
+
+        return num_windows
+
+    def _reshape_np_array(
+        self, array: NDArray[np.float_], window_size: int
+    ) -> NDArray[np.float_]:
+        """
+        Reshape a numpy array to a 2D array where each row represents a window.
+
+        There are three cases to consider
+
+        1. the array size is a multiple of window_size + period,
+           i.e. num_windows is integer and we can simply reshape
+        2. the array size is NOT a multiple of window_size + period and
+           it has length n * (window_size + period) + m, where m < window_size
+        3. the array size is NOT a multiple of window_size + period and
+           it has length n * (window_size + period) + m, where m >= window_size
```

Comment:
Can this happen?

Reply:
only case 1 can happen in the current implementation. But afair all three cases are covered in the benchmarks or the tests. I'll leave it for future work, when we remove the condition and add a comment.